src/webgpu/shader/execution/expression/call/builtin/determinant.cache.ts

```diff
@@ -1,60 +1,130 @@
+//   by the determinant of an (N-1)x(N-1) submatrix.
+//   The recurrence is:   T(N) = (N-1) + N + T(N-1), with T(1) = 0.
+//   The first N is for the adds and subtracts, the second N is for the multiplies,
+//   and the T(N-1) is the determinant of the submatrix.
+//   As N grows, this blows up quickly, even when accounting for reusing results.
+//   It's unlikely any GPU would actually calculate it this way.
+//
 // Accuracy for determinant is only defined for e, where e is an integer and
 // |e| < quadroot(2**21) [~38],
 // due to computational complexity of calculating the general solution for 4x4,
 // so custom matrices are used.
-//
-// Note: For 2x2 and 3x3 the limits are squareroot and cuberoot instead of
-// quadroot, but using the tighter 4x4 limits for all cases for simplicity.
-const kDeterminantValues = [-38, -10, -5, -1, 0, 1, 5, 10, 38];
 
-const kDeterminantMatrixValues = {
-  2: kDeterminantValues.map((f, idx) => [
-    [idx % 4 === 0 ? f : idx, idx % 4 === 1 ? f : -idx],
-    [idx % 4 === 2 ? f : -idx, idx % 4 === 3 ? f : idx],
-  ]),
-  3: kDeterminantValues.map((f, idx) => [
-    [idx % 9 === 0 ? f : idx, idx % 9 === 1 ? f : -idx, idx % 9 === 2 ? f : idx],
-    [idx % 9 === 3 ? f : -idx, idx % 9 === 4 ? f : idx, idx % 9 === 5 ? f : -idx],
-    [idx % 9 === 6 ? f : idx, idx % 9 === 7 ? f : -idx, idx % 9 === 8 ? f : idx],
-  ]),
-  4: kDeterminantValues.map((f, idx) => [
-    [
-      idx % 16 === 0 ? f : idx,
-      idx % 16 === 1 ? f : -idx,
-      idx % 16 === 2 ? f : idx,
-      idx % 16 === 3 ? f : -idx,
-    ],
-    [
-      idx % 16 === 4 ? f : -idx,
-      idx % 16 === 5 ? f : idx,
-      idx % 16 === 6 ? f : -idx,
-      idx % 16 === 7 ? f : idx,
-    ],
-    [
-      idx % 16 === 8 ? f : idx,
-      idx % 16 === 9 ? f : -idx,
-      idx % 16 === 10 ? f : idx,
-      idx % 16 === 11 ? f : -idx,
-    ],
-    [
-      idx % 16 === 12 ? f : -idx,
-      idx % 16 === 13 ? f : idx,
-      idx % 16 === 14 ? f : -idx,
-      idx % 16 === 15 ? f : idx,
-    ],
-  ]),
-};
+type Dim = 2 | 3 | 4;
+type FPWidth = 16 | 32;
+
+// Number of random matrices to test, per configuration.
+const numSamples = 20;
+
+// Returns a random element in a range suitable for a square matrix
+// that can have an accurately computed determinant.
+function randomMatrixEntry(p: PRNG, dim: Dim, fpwidth: FPWidth) {
+  // See above for the justification.
+  const rangeTable = {
+    16: {
+      2: 32,
+      3: 8,
+      4: 4,
+    },
+    32: {
+      2: 2896,
+      3: 161,
+      4: 38,
+    },
+  };
+  const N = rangeTable[fpwidth][dim];
+  // Centre the distribution roughtly at zero.
+  const balanced = p.uniformInt(N) - Math.floor(N / 2);
+  return balanced;
+}
+
+// Returns a random square matrix that should have an exactly computed
+// determinant for the given floating point width.
+function randomSquareMatrix(p: PRNG, dim: Dim, fpwidth: FPWidth) {
+  const result = Array(dim);
```

Comment:
Not a big fan using an implicit `any[]` here, since a) 'any' is basically an escape hatch out of TS's type system, and it really hides teh fact that this is actually an number[][].

As a spot fix, I think you can change this to being ` = new Array<number[]>(dim)` and down below use `result[r] = new Array<number>(dim)`

I think a more complete fix is to use the 2D Arraybuilder pattern from `unflatten2DArray` (in math.ts). Applied here it would look something like this:
```
  const result: number[][] = [...Array(dim)].map(_ => [...Array(dim)]);
  // Scale each element by a simple power of two. This should only affect
  // the exponent of the result.
  const multiplier = [1, 2, 0.25][p.uniformInt(3)];
  for (let c = 0; c < dim; c++) {
    for (let r = 0; r < dim; r++) {
      result[c][r] = multiplier * randomMatrixEntry(p, dim, fpwidth);
    }
  }
  return result;
```

Reply:
Heh. I missed the fact i wasn't constraining result's type very  much.
Thanks for the pattern suggestion. Will fix.